criu/cr-restore.c

```diff
@@ -3377,7 +3377,15 @@ static struct thread_creds_args *rst_prep_creds_args(CredsEntry *ce, unsigned lo
 	if (!args)
 		return ERR_PTR(-ENOMEM);
 
-	args->cap_last_cap = kdat.last_cap;
+	/* When restoring capability bounding set, don't attempt to mask out
+	 * caps which weren't known at the dumping side but are known at the
+	 * restore side - the dumped process couldn't have set that mask, and
+	 * the default is to not mask them out. Also in the opposite direction:
```

Comment:
>  the dumped process couldn't have set that mask, and the default is to not mask them out.

hmm, it looks more logical to drop all new caps from restored bounding sets, doesn't it?

Reply:
The idea is that restored processes should not be able to use new capabilities.

Reply:
The reasoning here is that if a checkpointing kernel knew more caps than restoring, then the extra caps can't be used nor disabled on the restoring side; if the restoring one knows more caps, then the extra caps coundn't have been disabled on the checkpointing side. (Note: there is no way to add caps to the bounding set.)

Reply:
>  if the restoring one knows more caps, then the extra caps coundn't have been disabled on the checkpointing side. (Note: there is no way to add caps to the bounding set.)

but the app doesn't rely on new caps and so they can be dropped from the bounding set without affecting the app behavior.

Reply:
Which app do you have in mind? If it calls a setuid (external machine-specific file) binary from time to time, it will break if the external binary needs those caps. Bounding set is not what modifies the C/Rd app itself but the processes it starts. (BTW, we use app-assisted C/R, so the app can delay the checkpoint until after some external command finishes if necessary.)

Reply:
> Which app do you have in mind? 

@osctobe runc (used by docker, podman, k8s, etc) starts containers with bounding sets that are specified in configs:
https://github.com/opencontainers/runtime-spec/blob/48415de180cf7d5168ca53a5aa27b6fcec8e4d81/specs-go/config.go#L103

> (BTW, we use app-assisted C/R, so the app can delay the checkpoint until after some external command finishes if necessary.)

I understand the problem that you are trying to solve, but here is the rule for such cases: in case of any uncertainty, we choose the variant that doesn't weaken security.

Here is an example. We have a process that is running without capabilities and with the empty bonding set. Then we migrate it to another node with a new kernel that has one new capability. Should the process be able to run any tools that  use this new capability. In general, the answer is no (it is the default case). In some cases like you described, the answer could be yes and we can help to figure out how to handle such cases.

Reply:
I've dug into the source of this patch and it looks it was fixing something else entirely. I redid the PR. We'll probably need more changes to allow unprivileged restore in some cases.